daft/execution/physical_plan.py

```diff
@@ -396,6 +399,628 @@ def broadcast_join(
+        .add_instruction(
+            instruction=execution_step.Project(
+                projection=ExpressionsProjection(
+                    [
+                        e.alias(right_name)
+                        for e, right_name in zip(left_on.to_column_expressions(), [e.name() for e in right_on])
+                    ]
+                ),
+            )
+        ).finalize_partition_task_single_output(stage_id=stage_id_reduce)
+    )
+    yield right_boundaries
+
+    # Wait for right-side boundaries to materialize.
+    while not right_boundaries.done():
+        logger.debug("sort blocked on completion of boundary partition: %s", right_boundaries)
+        yield None
+
+    # Sort both children using the combined boundaries.
+    sorted_plans: list[InProgressPhysicalPlan[PartitionT]] = []
+    for on, source_materializations, boundaries in [
+        (left_on, left_source_materializations, left_boundaries),
+        (right_on, right_source_materializations, right_boundaries),
+    ]:
+        # NOTE: We need to give reduce() an iter(list), since giving it a generator would result in lazy
+        # binding in this loop.
+        range_fanout_plan = [
+            PartitionTaskBuilder[PartitionT](
+                inputs=[boundaries.partition(), source.partition()],
+                partial_metadatas=[boundaries.partition_metadata(), source.partition_metadata()],
+                resource_request=ResourceRequest(
+                    memory_bytes=source.partition_metadata().size_bytes,
+                ),
+            ).add_instruction(
+                instruction=execution_step.FanoutRange[PartitionT](
+                    _num_outputs=num_partitions,
+                    sort_by=on,
+                    descending=descending,
+                ),
+            )
+            for source in consume_deque(source_materializations)
+        ]
+
+        # Execute a sorting reduce on it.
+        per_partition_bounds = _to_per_partition_bounds(boundaries.vpartition(), num_partitions)
+        sorted_plans.append(
+            reduce(
+                fanout_plan=iter(range_fanout_plan),
+                reduce_instructions=[
+                    execution_step.ReduceMergeAndSort(
+                        sort_by=on,
+                        descending=descending,
+                        bounds=per_part_boundaries,
+                    )
+                    for per_part_boundaries in per_partition_bounds
+                ],
+            )
+        )
+
+    left_sorted_plan, right_sorted_plan = sorted_plans
+
+    # Merge-join the two sorted sides of the join.
+    yield from merge_join_sorted(left_sorted_plan, right_sorted_plan, left_on, right_on, how, left_is_larger)
+
+
+def _to_per_partition_bounds(boundaries: MicroPartition, num_partitions: int) -> list[MicroPartition]:
+    boundaries_dict = boundaries.to_pydict()
+    return [
```

Comment:
this might break if `boundaries` is empty or all None. since it wont be able to infer types

Reply:
For both the empty `boundaries` case, which I suppose would arise if the dataframe has all-empty partitions, and the all `None` case, the result would be all `None` bounds for each partition bounds, which seems like the correct behavior to me.

As for not being able to infer types, we do currently treat `None` bounds explicitly in our boundaries comparator, where `None` indicates no bound, i.e. the most conservative interpretation. Do you seen an issue with those semantics?

Also, this should be covered by the all-null DataFrame-level tests, I think. https://github.com/Eventual-Inc/Daft/blob/b90299be064ed35facc5701dfe7da09a8e3af0f0/tests/dataframe/test_joins.py#L322-L350